cmd/red/proxy.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/segmentio/conf"
+	consul "github.com/segmentio/consul-go"
+	"github.com/segmentio/events"
+	eventslog "github.com/segmentio/events/log"
+	redis "github.com/segmentio/redis-go"
+)
+
+type proxyConfig struct {
+	Bind     string `conf:"bind"     help:"Address on which the proxy is listening for incoming connections." validate:"nonzero"`
+	Upstream string `conf:"upstream" help:"URL or comma-separated list of upstream servers."                  validate:"nonzero"`
+	Debug    bool   `conf:"debug"    help:"Enable debug mode."`
+}
+
+func proxy(args []string) (err error) {
+	config := proxyConfig{
+		Bind: ":6479",
+	}
+
+	conf.LoadWith(&config, conf.Loader{
+		Name: "red proxy",
+		Args: args,
+		Sources: []conf.Source{
+			conf.NewEnvSource("RED", os.Environ()...),
+		},
+	})
+
+	events.DefaultLogger.EnableDebug = config.Debug
+	events.DefaultLogger.EnableSource = config.Debug
+
+	defer func() {
+		if err == nil {
+			err = convertPanicToError(recover())
+		}
+	}()
+
+	lstn := makeListener(config.Bind)
+	server := makeProxyServer(config)
+
+	sigchan, sigstop := signals(syscall.SIGINT, syscall.SIGTERM)
+	defer sigstop()
+
+	go func() {
+		<-sigchan
+		sigstop()
+		server.Shutdown(context.Background())
+	}()
+
+	events.Log("listening on '%{address}s' for incoming connections", lstn.Addr())
+
+	if err = server.Serve(lstn); err == redis.ErrServerClosed {
+		err = nil
+	}
+
+	return
+}
+
+func makeListener(addr string) net.Listener {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		panic(err)
+	}
+	return l
+}
+
+func makeProxyServer(config proxyConfig) *redis.Server {
+	logger := eventslog.NewLogger("", 0, events.DefaultHandler)
+	return &redis.Server{
+		Handler:      makeReverseProxy(config, logger),
+		ReadTimeout:  30 * time.Second,
+		WriteTimeout: 30 * time.Second,
+		IdleTimeout:  90 * time.Second,
+		ErrorLog:     logger,
+	}
+}
+
+func makeReverseProxy(config proxyConfig, logger *log.Logger) redis.Handler {
+	return &redis.ReverseProxy{
+		Transport: makeTransport(),
+		Registry:  makeRegistry(config.Upstream),
+		ErrorLog:  logger,
+	}
+}
+
+func makeTransport() redis.RoundTripper {
+	// TODO: add metrics
+	return &redis.Transport{
+		ConnsPerHost: 4,
+		PingTimeout:  10 * time.Second,
+		PingInterval: 15 * time.Second,
+	}
+}
+
+func makeRegistry(upstream string) (registry redis.ServerRegistry) {
+	if strings.Index(upstream, "://") < 0 {
+		registry = makeStaticRegistry(upstream)
+	} else {
+		u, err := url.Parse(upstream)
+		if err != nil {
+			panic(err)
+		}
+
+		switch u.Scheme {
+		case "consul":
+			registry = makeConsulRegistry(u)
+
+		default:
+			panic("unsupported registry: " + u.Scheme)
```

Comment:
I generally agree, however within the code here which is mostly constructions of objects I adopted this pattern to make the code shorter and easier to follow by getting rid of the error checking and instead use an exception-like model based on panic.

The Go guidelines are that panics should not leave package boundaries to we're respecting this here.